cpp/dolfinx/io/checkpointing.cpp

```diff
@@ -0,0 +1,354 @@
+// Copyright (C) 2024 Abdullah Mujahid
+//
+// This file is part of DOLFINX (https://www.fenicsproject.org)
+//
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+#ifdef HAS_ADIOS2
+
+#include "checkpointing.h"
+#include <adios2.h>
+#include <basix/finite-element.h>
+#include <dolfinx/fem/CoordinateElement.h>
+#include <dolfinx/mesh/Mesh.h>
+#include <dolfinx/mesh/utils.h>
+#include <mpi.h>
+
+/// @file checkpointing.h
+/// @brief ADIOS2 based checkpointing
+namespace
+{
+std::map<basix::element::lagrange_variant, std::string> variant_to_string{
+    {basix::element::lagrange_variant::unset, "unset"},
+    {basix::element::lagrange_variant::equispaced, "equispaced"},
+    {basix::element::lagrange_variant::gll_warped, "gll_warped"},
+};
+
+std::map<std::string, basix::element::lagrange_variant> string_to_variant{
+    {"unset", basix::element::lagrange_variant::unset},
+    {"equispaced", basix::element::lagrange_variant::equispaced},
+    {"gll_warped", basix::element::lagrange_variant::gll_warped},
+};
+
+} // namespace
+
+namespace dolfinx::io::checkpointing
+{
+//-----------------------------------------------------------------------------
+template <std::floating_point T>
+void write_mesh(ADIOS2Wrapper& ADIOS2, dolfinx::mesh::Mesh<T>& mesh)
+{
+
+  const mesh::Geometry<T>& geometry = mesh.geometry();
+  std::shared_ptr<const mesh::Topology> topology = mesh.topology();
+
+  // Variables/attributes to save
+  std::int32_t dim = geometry.dim();
+  std::uint64_t num_vertices_global, offset, num_cells_global, cell_offset;
+  std::uint32_t num_vertices_local, num_cells_local, num_dofs_per_cell, degree;
+  std::string cell_type, lagrange_variant;
+  std::vector<std::int64_t> array_global;
+  std::vector<std::int32_t> offsets_global;
+
+  std::shared_ptr<const common::IndexMap> geom_imap;
+
+  // Vertices information
+  {
+    geom_imap = mesh.geometry().index_map();
+    num_vertices_global = geom_imap->size_global();
+    num_vertices_local = geom_imap->size_local();
+    offset = geom_imap->local_range()[0];
+  }
+
+  // Cells information
+  {
+    const std::shared_ptr<const common::IndexMap> topo_imap
+        = topology->index_map(dim);
+    num_cells_global = topo_imap->size_global();
+    num_cells_local = topo_imap->size_local();
+    cell_offset = topo_imap->local_range()[0];
+  }
+
+  // Coordinate element information
+  {
+    const fem::CoordinateElement<T>& cmap = mesh.geometry().cmap();
+    fem::ElementDofLayout geom_layout = cmap.create_dof_layout();
+    num_dofs_per_cell = geom_layout.num_entity_closure_dofs(dim);
```

Comment:
```suggestion
    num_dofs_per_cell = geom_layout.num_entity_closure_dofs(topology()->dim());
```
This should be the topological dimension.
Imagine a 2D manifold embedded in 3D space. The num entity closure dofs is still of dimension 2 (associated with the 2D cell), not 3.